ee/hogai/graph/taxonomy_agent/nodes.py

```diff
@@ -201,58 +319,84 @@ def _construct_messages(self, state: AssistantState) -> list[BaseMessage]:
+                    "id": tool_call_id,
+                    "name": action.tool,
+                    "partial_json": json.dumps(action.tool_input),
+                },
+            ]
+            if action.log and LOG_SEPARATOR in action.log:
+                _tool_call_id, thinking_content, thinking_signature = action.log.split(LOG_SEPARATOR)
+                assistant_content.insert(
+                    0,
+                    {
+                        "type": "thinking",
+                        "thinking": thinking_content,
+                        "signature": thinking_signature,
+                    },
+                )
+            conversation.append(
+                LangchainAssistantMessage(
+                    content=assistant_content,  # type: ignore
+                    tool_calls=[{"id": tool_call_id, "name": action.tool, "args": action.tool_input}],
+                )
+            )
+            conversation.append(LangchainToolMessage(content=output or "", tool_call_id=tool_call_id))
 
-    def _get_agent_scratchpad(self, scratchpad: list[tuple[AgentAction, str | None]]) -> str:
-        actions = []
-        for action, observation in scratchpad:
-            if observation is None:
-                continue
-            actions.append((action, observation))
-        return format_log_to_str(actions)
+        return conversation
 
 
-class TaxonomyAgentPlannerToolsNode(AssistantNode, ABC):
+class QueryPlannerToolsNode(AssistantNode, ABC):
     MAX_ITERATIONS = 16
     """
     Maximum number of iterations for the ReAct agent. After the limit is reached,
     the agent will terminate the conversation and return a message to the root node
     to request additional information.
     """
 
-    def _run_with_toolkit(
-        self, state: AssistantState, toolkit: TaxonomyAgentToolkit, config: Optional[RunnableConfig] = None
-    ) -> PartialAssistantState:
+    def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistantState:
+        toolkit = TaxonomyAgentToolkit(self._team)
         intermediate_steps = state.intermediate_steps or []
-        action, observation = intermediate_steps[-1]
+        action, _output = intermediate_steps[-1]
 
         input = None
         output = ""
 
         try:
-            input = TaxonomyAgentTool.model_validate({"name": action.tool, "arguments": action.tool_input}).root
+            if action.tool in [
+                "retrieve_entity_properties",
+                "retrieve_event_properties",
+                "retrieve_action_properties",
+                "ask_user_for_help",
+            ]:
+                if not isinstance(action.tool_input, dict):
+                    raise ValueError(f"Input for {action.tool} must be a dictionary, got {action.tool_input}")
+                arguments = next(iter(action.tool_input.values()))
+            else:
+                arguments = action.tool_input
+            input = TaxonomyAgentTool.model_validate(
```

Comment:
Shouldn't this also include the dynamic entity tools? So, `tools` and `_get_tools` should be dynamic too?